content/posts/2021/reana-0.7.4.md

```diff
@@ -0,0 +1,134 @@
+the cluster will respond with out of memory (OOM) error and the job will be terminated (OOMKilled).
+To avoid such an error and to make an efficient use of the available resources,
+it is recommended to specify a memory limit per workflow step.
+If a memory limit for a specific step is increased, OOM error will not be thrown and
+the workflow will be executed as expected. Also, a memory limit could be decreased
+in order to increment the chances of jobs being scheduled earlier.
+
+REANA 0.7.4 enables to set a memory limit for a workflow jobs by specifying
+`kubernetes_memory_limit` in the specification of each workflow step.
+Please refer to
+[Kubernetes official documentation](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)
+for expected memory values.
+The specifications respectively can be configured for Serial, Yadage,
+and CWL workflow engines.
+
+### Custom memory limit for Serial workflow
+
+We can set `kubernetes_memory_limit` in every step of workflow specification.
+
+
+```diff
+  ...
+  steps:
+    - name: gendata
+      environment: 'reanahub/reana-env-root6:6.18.04'
+      compute_backend: kubernetes
++     kubernetes_memory_limit: '8Gi'
+      commands:
+      - mkdir -p results && root -b -q 'code/gendata.C(${events},"${data}")'
+    - name: fitdata
+      environment: 'reanahub/reana-env-root6:6.18.04'
+      compute_backend: kubernetes
++     kubernetes_memory_limit: '4Gi'
+      commands:
+      - root -b -q 'code/fitdata.C("${data}","${plot}")'
+```
+
+
+### Custom memory limit for Yadage workflow
+
+We can set `kubernetes_memory_limit` in every stage step under `environment.resources`.
+
+```diff
+  ...
+  stages:
+    - name: gendata
+      dependencies: [init]
+      scheduler:
+        scheduler_type: 'singlestep-stage'
+        parameters:
+          events: {step: init, output: events}
+          gendata: {step: init, output: gendata}
+          outfilename: '{workdir}/data.root'
+        step:
+          process:
+            process_type: 'interpolated-script-cmd'
+            script: root -b -q '{gendata}({events},"{outfilename}")'
+          publisher:
+            publisher_type: 'frompar-pub'
+            outputmap:
+              data: outfilename
+          environment:
+            environment_type: 'docker-encapsulated'
+            image: 'reanahub/reana-env-root6'
+            imagetag: '6.18.04'
+            resources:
+              - compute_backend: kubernetes
++             - kubernetes_memory_limit: '8Gi'
+```
+### Custom memory limit for CWL workflow
+
+We can set `kubernetes_memory_limit` in every step under `hints.reana`.
+
+```diff
+  ...
+  steps:
+    gendata:
+      run: gendata.cwl
+      in:
+        gendata_tool: gendata_tool
+        events: events
+      hints:
+        reana:
+          compute_backend: kubernetes
++         kubernetes_memory_limit: '8Gi'
+      out: [data]
+```
+
+### Custom memory limit administration
```

Comment:
Cluster-wide defaults

REANA cluster administrators can configure the default memory limit for user job containers by setting kubernetes_jobs_memory_limit Helm value to the desired value such as "8Gi". Please note that the best value to use would  depend both on the typical user workflows and on the parameters of the cluster nodes, most notably the available memory on the nodes. Typically a few GiBs should be left reserved for the Kubernetes system, so for the cluster nodes of 16 GiB the maximum limit should be around 12 GiB.

REANA cluster administrators can also set the maximum custom memory limit value that users would be able to ask for in their custom job resource clauses. The value could be changed via kubernetes_jobs_max_user_memory_limit Helm value during cluster deployment.  If a user asks for more memory, the workflow would not be started and an error will be signalled to the user.

Please refer to the official Kubernetes documentation to know more about the container memory limits and the expected value format.